File: vol2/ch18/LeetCode127-1.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int n=wordList.size();
        unordered_set<string> dict;     //字典用dict表示
        unordered_set<string> visited;  //是否访问过
        for(auto s:wordList) dict.insert(s);
        if(!dict.count(endWord)) return 0;
        queue<string> qu;
        qu.push(beginWord);
        visited.insert(beginWord);
        int ans=0;
        while(!qu.empty()) {
            ans++;
            int cnt=qu.size();
            for(int k=0;k<cnt;k++) {
                string str=qu.front(); qu.pop();    //出队str
                for(int i=0;i<str.size();i++) {     //对str每个字符进行改变
                    string copystr=str;
                    for(int j=0;j<26;j++) {
                        copystr[i]='a'+j;
                        if(copystr==endWord)
                            return ans+1;
                        if(dict.count(copystr)==0)
                            continue;               //copystr不在字典中时跳过
                        if(visited.count(copystr))
                            continue;               //copystr已访问过时跳过
                        qu.push(copystr);
                        visited.insert(copystr);
                    }
                }
            }
        }
        return 0;
    }
};

File: vol2/ch18/LeetCode127-1.py
class Solution:
    def ladderLength(self,beginWord:str,endWord:str,wordList:List[str])->int:
        dict=set()     #字典用dict表示
        visited=set()  #是否访问过
        for s in wordList:dict.add(s)
        if endWord not in dict:return 0
        qu=deque()
        qu.append(beginWord)
        visited.add(beginWord)
        ans=0
        while qu:
            ans+=1
            cnt=len(qu)
            for k in range(0,cnt):
                str=qu.popleft()        #出队str
                for i in range(0,len(str)): #对str每个字符进行改变
                    for j in range(0,26):
                        tmp=list(str)       #将str转换为列表
                        tmp[i]=chr(ord('a')+j); #修改tmp[i]
                        copystr=''.join(tmp)        #将tmp转换为字符串copystr
                        if copystr==endWord:
                            return ans+1
                        if copystr not in dict:
                            continue               #copystr不在字典中时跳过
                        if copystr in visited:
                            continue               #copystr已访问过时跳过
                        qu.append(copystr)
                        visited.add(copystr)
        return 0
